solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java

```diff
@@ -0,0 +1,250 @@
+              withMethod(GET)
+              .build().process(cluster.getSolrClient()),
+          Utils.makeMap(
+              ":result:packages:mypkg[0]:version", "1.0",
+              ":result:packages:mypkg[0]:files[0]", FILE1
+          ));
+
+      PluginMeta plugin = new PluginMeta();
+      plugin.name = "myplugin";
+      plugin.klass = "mypkg:org.apache.solr.handler.MyPlugin";
+      plugin.version = add.version;
+      V2Request req1 = new V2Request.Builder("/cluster/plugin")
+          .forceV2(true)
+          .withMethod(POST)
+          .withPayload(singletonMap("add", plugin))
+          .build();
+      req1.process(cluster.getSolrClient());
+      TestDistribPackageStore.assertResponseValues(10,
+          () -> new V2Request.Builder("/cluster/plugin").
+              withMethod(GET)
+              .build().process(cluster.getSolrClient()),
+          ImmutableMap.of(
+              "/plugin/myplugin/class", plugin.klass,
+              "/plugin/myplugin/version", plugin.version
+          ));
+      Callable<NavigableObject> invokePlugin = () -> new V2Request.Builder("/plugin/my/path")
+          .forceV2(true)
+          .withMethod(GET)
+          .build().process(cluster.getSolrClient());
+      TestDistribPackageStore.assertResponseValues(10,
+          invokePlugin,
+          ImmutableMap.of("/myplugin.version", "1.0"));
+
+      add.version = "2.0";
+      add.files = List.of(FILE2);
+      addPkgVersionReq.process(cluster.getSolrClient());
+
+      plugin.version = add.version;
+      new V2Request.Builder("/cluster/plugin")
+          .forceV2(true)
+          .withMethod(POST)
+          .withPayload(singletonMap("update", plugin))
+          .build()
+      .process(cluster.getSolrClient());
+
+      TestDistribPackageStore.assertResponseValues(10,
+          () -> new V2Request.Builder("/cluster/plugin").
+              withMethod(GET)
+              .build().process(cluster.getSolrClient()),
+          ImmutableMap.of(
+              "/plugin/myplugin/class", plugin.klass,
+              "/plugin/myplugin/version", "2.0"
+          ));
+      TestDistribPackageStore.assertResponseValues(10,
+          invokePlugin,
+          ImmutableMap.of("/myplugin.version", "2.0"));
+    } finally {
+      cluster.shutdown();
+    }
+  }
+
+  public static class C1 {
+
+  }
+
+  @EndPoint(
+      method = GET,
+      path = "/plugin/my/plugin",
+      permission = PermissionNameProvider.Name.COLL_READ_PERM)
+  public class C2 {
+
+
+  }
+
+  @EndPoint(
+      method = GET,
+      path = "/plugin/my/plugin",
+      permission = PermissionNameProvider.Name.COLL_READ_PERM)
+  public static class C3 {
+    @Command
+    public void read(SolrQueryRequest req, SolrQueryResponse rsp) {
+      rsp.add("testkey", "testval");
+    }
+
+  }
+
+
+  private void expectError(V2Request req, SolrClient client, String errPath, String expectErrorMsg) throws IOException, SolrServerException {
+    try {
```

Comment:
use `expectThrows` here.

Reply:
`expectThrows` does not work here

Reply:
How so?  It returns the exception of the expected type and then you can inspect it as you are doing now.

Reply:
fixed